### PR TITLE
Shm ref count

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

File: core/arch/arm/tee/entry_std.c [hunk 1]
```diff
@@ -174,17 +183,16 @@ static TEE_Result copy_in_params(const struct optee_msg_param *params,
 	return TEE_SUCCESS;
 }
 
-static void cleanup_params(const struct optee_msg_param *params,
-			   const uint64_t *saved_attr,
-			   uint32_t num_params)
+static void cleanup_shm_refs(const uint64_t *saved_attr,
```
Comment (lorc) [inline core/arch/arm/tee/entry_std.c]: Hi, @jenswi-linaro, 

this change is not covered in the commit message. And I think, it should be moved into separate commit.

Comment (jenswi-linaro) [reply core/arch/arm/tee/entry_std.c]: This is a central part in the commit, especially the previously saved shm_ref should be used instead of reading it again.
Sure, I could delay renaming the function but that is bordering on splitting hairs.

File: core/arch/arm/mm/mobj.c [hunk 2]
```diff
@@ -461,21 +487,118 @@ struct mobj *mobj_reg_shm_alloc(paddr_t *pages, size_t num_pages,
...
+		 * we're in trouble.
+		 */
+		if (!refcount_inc(&r->refcount))
+			panic();
 	}
+
 	cpu_spin_unlock_xrestore(&reg_shm_slist_lock, exceptions);
-	return NULL;
+
+	return &r->mobj;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj.c]: Well spotted, I'll fix

File: core/arch/arm/mm/mobj.c [hunk 3]
```diff
@@ -461,21 +487,118 @@ struct mobj *mobj_reg_shm_alloc(paddr_t *pages, size_t num_pages,
...
+{
+	uint32_t exceptions = cpu_spin_lock_xsave(&reg_shm_slist_lock);
+	struct mobj_reg_shm *r = reg_shm_find_unlocked(cookie);
+
+	/*
+	 * A put is supposed to match a get, if the object isn't found
+	 * it's out of sync somehow. The counter is supposed to be larger
+	 * than 1, if it isn't we're in trouble.
+	 */
+	if (!r || refcount_dec(&r->refcount))
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj.c]: Not it will not panic OP-TEE as we're not supposed to pass the get that has to succeed before this.
I'll update the comment.

File: core/arch/arm/mm/mobj.c [hunk 4]
```diff
@@ -461,21 +487,118 @@ struct mobj *mobj_reg_shm_alloc(paddr_t *pages, size_t num_pages,
...
+{
+	TEE_Result res = TEE_ERROR_BAD_PARAMETERS;
+	uint32_t exceptions = cpu_spin_lock_xsave(&reg_shm_slist_lock);
+	struct mobj_reg_shm *r = reg_shm_find_unlocked(cookie);
+
+	if (!r)
+		goto out;
+
+	res = TEE_ERROR_BUSY;
+	if (refcount_val(&r->refcount) == 1) {
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj.c]: Correct, it's impossible due to the spinlock.

File: core/arch/arm/tee/entry_std.c [hunk 5]
```diff
@@ -142,6 +142,7 @@ static TEE_Result copy_in_params(const struct optee_msg_param *params,
 		switch (attr) {
 		case OPTEE_MSG_ATTR_TYPE_NONE:
 			pt[n] = TEE_PARAM_TYPE_NONE;
+			memset(&ta_param->u[n], 0, sizeof(ta_param->u[n]));
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/tee/entry_std.c]: Sorry, this isn't supposed to be here at all. I got carried over during a rebase by accident. I'll remove it.

File: core/arch/arm/mm/mobj.c [hunk 6]
```diff
@@ -4,10 +4,11 @@
  */
 
 #include <assert.h>
-#include <keep.h>
 #include <initcall.h>
+#include <keep.h>
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/mobj.c]: OK, I'll back this out

File: core/arch/arm/tee/entry_std.c (hunk 1 above)
Comment (lorc) [reply core/arch/arm/tee/entry_std.c]: I'm talking not about this particular change, but about whole entry_std.c diff.

At least you can mention this explicitly in the commit message. Because right now this patch does two things:

 * it is adding reference counter to `mobj_reg_shm`
 * and it alters `entry_std.c` to make proper use of this reference counter

So, it will be great to see both changes covered in the commit message.

**Comment by lorc** [comment]
With the commit message for "core: reference count registered shared memory" updated:

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

File: core/arch/arm/tee/entry_std.c (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/tee/entry_std.c]: Aha, now I get it. I'll split it.

Comment (jenswi-linaro) [reply core/arch/arm/tee/entry_std.c]: Hmm, it has to go in the same commit. I'll expand the commit message as suggested instead.

**Comment by jenswi-linaro** [comment]
Squashed and tag applied
